filesystem support for exclude and include filters

marchenko1985 opened the pull request:
Closes: #536 and technically closes: #865

What has been done:

- I have moved the git filter creation a little bit lower into its switch case
- And created similar filter creation in the filesystem switch case
- Because of that go asked me to put `err :=` below

For filesystem engine, itself did not want to touch it so just added one more "method" `WithFilter()` so it can be used where needed

In the same way, as in the git source, we are checking if the filter is not nil and the current file passes it

---

Offtopic: while trying to figure out how to do it found one more crappy thing in the filter itself - it does not remove empty lines from the exclusion file, and it may lead to scenarios when by mistake we may exclude all files with space in the name

---

Tests: did not find any tests related to the filter in the git source so have nothing to copy from, and because I'm golang noob all that channel stuff just blows my brain so can not handle that one 🤷‍♂️

marchenko1985 committed:
fix filter issue - empty lines should be ignored

marchenko1985 committed:
filesystem support for filter exclude

ahrav commented on main.go:
```diff
@@ -174,15 +174,14 @@ func run(state overseer.State) {
-	}
-
 	var repoPath string
 	var remote bool
```
nit: If we declare `var err error` here, within the switch we can use regular assignment `=` vs `:=`

ahrav reviewed (approved):
Thanks a lot @mac2000 this looks like a great addition. Really appreciate it 😄 

Looks like there is a conflict in main, otherwise LGTM!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/trufflesecurity/trufflehog?pullRequest=881) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/trufflesecurity/trufflehog?pullRequest=881) before we can accept your contribution.<br/>**12** out of **13** committers have signed the CLA.<br/><br/>:white_check_mark: mac2000<br/>:white_check_mark: mcastorina<br/>:white_check_mark: dustin-decker<br/>:white_check_mark: ahrav<br/>:white_check_mark: jjasghar<br/>:white_check_mark: ankushgoel27<br/>:white_check_mark: kstilwell<br/>:white_check_mark: bill-rich<br/>:white_check_mark: jcastilloixl<br/>:white_check_mark: jesslam948<br/>:white_check_mark: trufflesteeeve<br/>:white_check_mark: jjsaunier<br/>:x: dependabot[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/trufflesecurity/trufflehog?pullRequest=881) it.</sub>

ahrav committed:
Use pointer to type. (#926)

ahrav committed:
Use pointer to type. (#926)

mcastorina committed:
Return an error from ReadToMax when it panics (#925)

mcastorina committed:
Add warning about using trufflehog as a library (#928)

dustin-decker committed:
Enable skipping of particular key IDs (#930)

* Enable skipping of particular key IDs

* update test

dustin-decker committed:
add LDAP detector (#896)

thiagola92 committed:
Fix error message (#933)

dustin-decker committed:
Skip some FTP FPs (#929)

jesslam948 committed:
add rambbitmq detector (#936)

* add rambbitmq detector

* use fixed length redaction

Co-authored-by: Dustin Decker <dustin@trufflesec.com>

ahrav committed:
Update slack webhook detector string check (#932)

* Update slack webhook detector check to text.

* remove redunant slashes.

mcastorina committed:
Add custom detectors configuration parsing (#927)

* Add custom_detectors proto

* Generate proto code

* Create custom_detectors package

Also create protoyaml package to test YAML unmarshalling the
configuration.

* Simplify custom_detectors proto by removing connection

* Generate proto code

* Update custom_detectors parsing tests

dustin-decker committed:
S3 scanner improvements (#938)

bill-rich committed:
Add newlines when file is split (#937)

0xflotus committed:
fix: small errors (#917)

bill-rich committed:
Close files to clean up tmp files (#940)

bill-rich committed:
Use new diskbufferreader version (#941)

bill-rich committed:
Add logger to context (#947)

* Add logger to context

* Fatal on no org

mcastorina committed:
Add test for configuring custom regex with webhook verification (#946)

mcastorina committed:
Add CustomRegex validation (#939)

* Add validation skeleton

* Add custom detector validation with tests

* Validate and test regex vars

* Implement RegexVarString

* Use RegexVarString for validating regex variables

* Add numerics to the regex variable matching

Co-authored-by: hxnyk <8292703+hxnyk@users.noreply.github.com>

dxa4481 committed:
Update README.md (#952)

bill-rich committed:
Integration test fixes (#956)

* Adjust repo count for new app

* Fix chunk test count

bill-rich committed:
Change chunker test source (#959)

* Change chunker test source

* Emit chunk if the size isn't 0

bill-rich committed:
Don't scan the --since-commit target (#960)

bill-rich committed:
Pre-commit not GH action PR (#961)

bill-rich committed:
Export line number code (#962)

pulkitanz committed:
Update Adding_Detectors_external.md (#957)

dependabot[bot] committed:
Bump github.com/go-sql-driver/mysql from 1.6.0 to 1.7.0 (#954)

Bumps [github.com/go-sql-driver/mysql](https://github.com/go-sql-driver/mysql) from 1.6.0 to 1.7.0.
- [Release notes](https://github.com/go-sql-driver/mysql/releases)
- [Changelog](https://github.com/go-sql-driver/mysql/blob/master/CHANGELOG.md)
- [Commits](https://github.com/go-sql-driver/mysql/compare/v1.6.0...v1.7.0)

---
updated-dependencies:
- dependency-name: github.com/go-sql-driver/mysql
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump go.uber.org/zap from 1.23.0 to 1.24.0 (#955)

Bumps [go.uber.org/zap](https://github.com/uber-go/zap) from 1.23.0 to 1.24.0.
- [Release notes](https://github.com/uber-go/zap/releases)
- [Changelog](https://github.com/uber-go/zap/blob/master/CHANGELOG.md)
- [Commits](https://github.com/uber-go/zap/compare/v1.23.0...v1.24.0)

---
updated-dependencies:
- dependency-name: go.uber.org/zap
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

ahrav committed:
Only scan org with --org flag. (#931)

dustin-decker committed:
Support globbing with ignore repos (#967)

dependabot[bot] committed:
Bump github.com/xanzy/go-gitlab from 0.74.0 to 0.76.0 (#934)

Bumps [github.com/xanzy/go-gitlab](https://github.com/xanzy/go-gitlab) from 0.74.0 to 0.76.0.
- [Release notes](https://github.com/xanzy/go-gitlab/releases)
- [Changelog](https://github.com/xanzy/go-gitlab/blob/master/releases_test.go)
- [Commits](https://github.com/xanzy/go-gitlab/compare/v0.74.0...v0.76.0)

---
updated-dependencies:
- dependency-name: github.com/xanzy/go-gitlab
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

bill-rich committed:
Bump github.com/bill-rich/disk-buffer-reader from v0.1.6 to v0.1.7 (#970)

dependabot[bot] committed:
Bump go.mongodb.org/mongo-driver from 1.11.0 to 1.11.1 (#971)

Bumps [go.mongodb.org/mongo-driver](https://github.com/mongodb/mongo-go-driver) from 1.11.0 to 1.11.1.
- [Release notes](https://github.com/mongodb/mongo-go-driver/releases)
- [Commits](https://github.com/mongodb/mongo-go-driver/compare/v1.11.0...v1.11.1)

---
updated-dependencies:
- dependency-name: go.mongodb.org/mongo-driver
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/getsentry/sentry-go from 0.15.0 to 0.16.0 (#973)

Bumps [github.com/getsentry/sentry-go](https://github.com/getsentry/sentry-go) from 0.15.0 to 0.16.0.
- [Release notes](https://github.com/getsentry/sentry-go/releases)
- [Changelog](https://github.com/getsentry/sentry-go/blob/master/CHANGELOG.md)
- [Commits](https://github.com/getsentry/sentry-go/compare/v0.15.0...v0.16.0)

---
updated-dependencies:
- dependency-name: github.com/getsentry/sentry-go
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

ahrav committed:
[bug] - Handle error when scanning s3 bucket. (#969)

* Handle error when scanning s# bucket.

* move wait outside loop.

* Add logging.

* revert changes.

* remove.

* revert.

dependabot[bot] committed:
Bump github.com/go-git/go-git/v5 from 5.4.2 to 5.5.1 (#972)

Bumps [github.com/go-git/go-git/v5](https://github.com/go-git/go-git) from 5.4.2 to 5.5.1.
- [Release notes](https://github.com/go-git/go-git/releases)
- [Commits](https://github.com/go-git/go-git/compare/v5.4.2...v5.5.1)

---
updated-dependencies:
- dependency-name: github.com/go-git/go-git/v5
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/envoyproxy/protoc-gen-validate from 0.6.13 to 0.9.1 (#963)

Bumps [github.com/envoyproxy/protoc-gen-validate](https://github.com/envoyproxy/protoc-gen-validate) from 0.6.13 to 0.9.1.
- [Release notes](https://github.com/envoyproxy/protoc-gen-validate/releases)
- [Changelog](https://github.com/bufbuild/protoc-gen-validate/blob/main/.goreleaser.yaml)
- [Commits](https://github.com/envoyproxy/protoc-gen-validate/compare/v0.6.13...v0.9.1)

---
updated-dependencies:
- dependency-name: github.com/envoyproxy/protoc-gen-validate
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mcastorina committed:
Add more logging for git sources (#974)

bill-rich committed:
Add s3 object count to trace logs (#975)

* Add s3 object count to trace logs

* fix debug level

mcastorina committed:
Implement CustomRegex detector (#950)

* Remove verifying successRanges because it is unused in webhook

* Move custom_detectors validation code into its own file

* Initial implementation of custom regex detector

Secret verification is done via webhook.

* Add CustomRegex detector type

* Add upper bound to permutation

* Return early if the context is canceled

* Add headers from configuration

* Add detector name as a key in the JSON body

* Implement faster algorithm for productIndices

goncalossilva committed:
Use Todoist's REST API v2 (#978)

v1 was deprecated on December 5, 2022.

ahrav committed:
Allow using a glob for include list. (#977)

* Allow using a glob for include list.

* Update command flag.

* Make comment more clear.

* update comment.

* Allow scanning repo and org at the same time.

dependabot[bot] committed:
Bump github.com/xanzy/go-gitlab from 0.76.0 to 0.77.0 (#981)

Bumps [github.com/xanzy/go-gitlab](https://github.com/xanzy/go-gitlab) from 0.76.0 to 0.77.0.
- [Release notes](https://github.com/xanzy/go-gitlab/releases)
- [Changelog](https://github.com/xanzy/go-gitlab/blob/master/releases_test.go)
- [Commits](https://github.com/xanzy/go-gitlab/compare/v0.76.0...v0.77.0)

---
updated-dependencies:
- dependency-name: github.com/xanzy/go-gitlab
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump golang.org/x/crypto from 0.3.0 to 0.4.0 (#982)

Bumps [golang.org/x/crypto](https://github.com/golang/crypto) from 0.3.0 to 0.4.0.
- [Release notes](https://github.com/golang/crypto/releases)
- [Commits](https://github.com/golang/crypto/compare/v0.3.0...v0.4.0)

---
updated-dependencies:
- dependency-name: golang.org/x/crypto
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mcastorina committed:
Add configuration parsing and custom detectors to engine (#968)

* Add configuration parsing for custom detectors

* Error on empty filename

mcastorina committed:
Add custom regex detector docs (#983)

mcastorina committed:
Remove custom log leveler (#985)

Instead of manually using a log leveler, rely on the global one defined
in the `log` package.

ahrav committed:
add proto that was missing. (#986)

yilmi committed:
Removing Debug version Println to logrus debug (#993)

ahrav committed:
Make GA action default base an empty string. (#996)

dustin-decker committed:
Add Circle CI source (#997)

* Add Circle CI source

* remove SHA1 line

* remove trim

ahrav committed:
Allow for default value to be used. (#999)

pulkitanz committed:
Fix GitUrl Return (#987)

Co-authored-by: ahrav <ahravdutta02@gmail.com>

dependabot[bot] committed:
Bump goreleaser/goreleaser-action from 3 to 4 (#980)

Bumps [goreleaser/goreleaser-action](https://github.com/goreleaser/goreleaser-action) from 3 to 4.
- [Release notes](https://github.com/goreleaser/goreleaser-action/releases)
- [Commits](https://github.com/goreleaser/goreleaser-action/compare/v3...v4)

---
updated-dependencies:
- dependency-name: goreleaser/goreleaser-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

yilmi committed:
Switch to retryableHttpClient for GitHub AuthN API Client + More Logs (#995)

* Adding missing flags to Readme

* Use retryableHttpClient by default for GitHub

* Adding repoUrl for scanning time log

* Use WithField instead of WithFields

* Updating README with lasted --help output

dependabot[bot] committed:
Bump cloud.google.com/go/secretmanager from 1.9.0 to 1.10.0 (#1006)

Bumps [cloud.google.com/go/secretmanager](https://github.com/googleapis/google-cloud-go) from 1.9.0 to 1.10.0.
- [Release notes](https://github.com/googleapis/google-cloud-go/releases)
- [Changelog](https://github.com/googleapis/google-cloud-go/blob/main/documentai/CHANGES.md)
- [Commits](https://github.com/googleapis/google-cloud-go/compare/dlp/v1.9.0...asset/v1.10.0)

---
updated-dependencies:
- dependency-name: cloud.google.com/go/secretmanager
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/go-git/go-git/v5 from 5.5.1 to 5.5.2 (#1007)

Bumps [github.com/go-git/go-git/v5](https://github.com/go-git/go-git) from 5.5.1 to 5.5.2.
- [Release notes](https://github.com/go-git/go-git/releases)
- [Commits](https://github.com/go-git/go-git/compare/v5.5.1...v5.5.2)

---
updated-dependencies:
- dependency-name: github.com/go-git/go-git/v5
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/hashicorp/go-retryablehttp from 0.7.1 to 0.7.2 (#1008)

Bumps [github.com/hashicorp/go-retryablehttp](https://github.com/hashicorp/go-retryablehttp) from 0.7.1 to 0.7.2.
- [Release notes](https://github.com/hashicorp/go-retryablehttp/releases)
- [Commits](https://github.com/hashicorp/go-retryablehttp/compare/v0.7.1...v0.7.2)

---
updated-dependencies:
- dependency-name: github.com/hashicorp/go-retryablehttp
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump golang.org/x/crypto from 0.4.0 to 0.5.0 (#1009)

Bumps [golang.org/x/crypto](https://github.com/golang/crypto) from 0.4.0 to 0.5.0.
- [Release notes](https://github.com/golang/crypto/releases)
- [Commits](https://github.com/golang/crypto/compare/v0.4.0...v0.5.0)

---
updated-dependencies:
- dependency-name: golang.org/x/crypto
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

ahrav committed:
Handle invalid regex for custom detector. (#1005)

* Handle invalid regex for custom detector.

* Add comment highlighting invalid regex.

mcastorina committed:
Capture callstack of canceled contexts (#979)

mcastorina committed:
Validate custom regular expressions on detector initialization (#1010)

* Validate custom regular expressions on detector initialization

* Add regex name to error message

clarkedb committed:
Revert "Allow for default value to be used. (#999)" (#1004)

This reverts commit ee6817ad85939d2b06ad6374eab3db4333fd8d46.

bill-rich committed:
Copy metadata for line number aware sources (#1011)

* Copy metadata for line number aware sources

* Improve style

bill-rich committed:
Rename and export isGitSource (#1016)

ahrav committed:
Update entrypoint (#1013)

* Update entrypoint.

* update comment.

* Call each extra arg on its own.

* Update loop.

* Update extra args.

* Use miccah's magic script.

* Fix path to bash.

* update entrypoint.

* Add bash to Dockerfile.

* update goreleaser dockerfile.

clonsdale-canva committed:
Use access-token endpoint for validity check (#991)

mcastorina committed:
Record timestamp when a context was cancelled (#1018)

dustin-decker committed:
remove logger from retryable client, it is not respecting loglevels (#1020)

ahrav committed:
[chore] - Small cleanup of CircleCi source (#1028)

* Small cleanup of CircleCi source.

* address comments.

* Add context to methods as first param.

ahrav committed:
Add concurrency to CircleCi source (#1029)

* Small cleanup of CircleCi source.

* Add concurrency to circleci.

* merge w/ cleanup branch.

* Rdefine loop var.

* Delete github.go

* reverge file delete.

* Add debug log for scan errors.

* make collecting scanned errors thread safe.

* pre-allocate errors slice.

dependabot[bot] committed:
Bump github.com/getsentry/sentry-go from 0.16.0 to 0.17.0 (#1022)

Bumps [github.com/getsentry/sentry-go](https://github.com/getsentry/sentry-go) from 0.16.0 to 0.17.0.
- [Release notes](https://github.com/getsentry/sentry-go/releases)
- [Changelog](https://github.com/getsentry/sentry-go/blob/master/CHANGELOG.md)
- [Commits](https://github.com/getsentry/sentry-go/compare/v0.16.0...v0.17.0)

---
updated-dependencies:
- dependency-name: github.com/getsentry/sentry-go
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/xanzy/go-gitlab from 0.77.0 to 0.78.0 (#1024)

Bumps [github.com/xanzy/go-gitlab](https://github.com/xanzy/go-gitlab) from 0.77.0 to 0.78.0.
- [Release notes](https://github.com/xanzy/go-gitlab/releases)
- [Changelog](https://github.com/xanzy/go-gitlab/blob/master/releases_test.go)
- [Commits](https://github.com/xanzy/go-gitlab/compare/v0.77.0...v0.78.0)

---
updated-dependencies:
- dependency-name: github.com/xanzy/go-gitlab
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump github.com/sergi/go-diff from 1.2.0 to 1.3.1 (#1023)

Bumps [github.com/sergi/go-diff](https://github.com/sergi/go-diff) from 1.2.0 to 1.3.1.
- [Release notes](https://github.com/sergi/go-diff/releases)
- [Commits](https://github.com/sergi/go-diff/compare/v1.2.0...v1.3.1)

---
updated-dependencies:
- dependency-name: github.com/sergi/go-diff
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

0x1 committed:
Updated stdout to print results in alphabetical order for consistent output (#1032)

mac2000 commented:
Hm, sorry, it was my fault, after messing with the GitHub sync of branches, I broke everything completely 🤷‍♂️

As a result figured out how to reset everything to upstream and cherry pick changes, otherwise my pull request was a mess of all changes from a PR creation

@ahrav if there is a chance to not wait for next conflict will be nice to merge it, seems like other similar changes were merged already (e.g. circleci flag for path exclusion)

> PS: I hope that GitHub will figure out what I'm trying to do, so no need to create another pull request, if I did everything correct, after reopening it we should see desired changes only

mac2000 commented:
@dickc-sg I broke it completely 🤷‍♂️ 

Looking forward to second attempt #1033

@ahrav if you still here will be nice to have you approve there as well so it might be merged

mac2000 commented:
@dickc-sg fyi seems like we have made it 💪

thanks to @dustin-decker 👍 #1033 pull request was approved and merged
